optaplanner-core/src/main/java/org/optaplanner/core/impl/util/ConsecutiveSetTree.java

```diff
@@ -0,0 +1,314 @@
+        }).collect(Collectors.toList());
+    }
+
+    public List<D> getConsecutiveLengths() {
+        return getConsecutiveSequences().stream()
+                .map(sequence -> differenceFunction.apply(indexFunction.apply(sequence.getItems().first()),
+                        indexFunction.apply(sequence.getItems().last())))
+                .collect(Collectors.toList());
+    }
+
+    public Optional<D> getBreakBefore(Sequence<T> sequence) {
+        T startItem = sequence.getItems().first();
+        T prevStartItem = startItemToSequence.lowerKey(startItem);
+        if (prevStartItem == null) {
+            return Optional.empty();
+        }
+        return Optional
+                .of(differenceFunction.apply(indexFunction.apply(startItemToSequence.get(prevStartItem).getItems().last()),
+                        indexFunction.apply(startItem)));
+    }
+
+    public Optional<D> getBreakAfter(Sequence<T> sequence) {
+        T endItem = sequence.getItems().last();
+        T nextStartItem = startItemToSequence.higherKey(endItem);
+        if (nextStartItem == null) {
+            return Optional.empty();
+        }
+        return Optional.of(differenceFunction.apply(indexFunction.apply(endItem),
+                indexFunction.apply(nextStartItem)));
+    }
+
+    public ConsecutiveData<T, D> getConsecutiveData() {
+        return consecutiveData;
+    }
+
+    public I getEndIndex(T key) {
+        return indexFunction.apply(startItemToSequence.get(key).getItems().last());
+    }
+
+    private boolean isSecondSuccessorOfFirst(I first, I second) {
+        D difference = differenceFunction.apply(second, first);
+        return !(difference.compareTo(maxDifference) > 0 || difference.compareTo(zeroDifference) < 0);
+    }
+
+    @Override
+    public boolean add(T item) {
+        T firstBeforeItem = startItemToSequence.floorKey(item);
+        I itemIndex = indexFunction.apply(item);
+        if (firstBeforeItem != null) {
+            I endIndex = getEndIndex(firstBeforeItem);
+            if (itemIndex.compareTo(endIndex) <= 0) {
+                // Item is already in the bag; increase it count
+                startItemToSequence.get(firstBeforeItem).add(item);
+            } else {
+                // Item is outside the bag
+                T firstAfterItem = startItemToSequence.higherKey(item);
+                if (firstAfterItem != null) {
+                    I afterStartIndex = indexFunction.apply(firstAfterItem);
+                    if (isSecondSuccessorOfFirst(itemIndex, endIndex)) {
+                        // We need to extend the first bag
+                        Sequence<T> prevBag = startItemToSequence.get(firstBeforeItem);
+                        if (isSecondSuccessorOfFirst(afterStartIndex, itemIndex)) {
+                            // We need to merge the two bags
+                            Sequence<T> afterBag = startItemToSequence.remove(firstAfterItem);
+                            prevBag.putAll(afterBag);
```

Comment:
The additions and removals from these bags are going to be where this code spends the most time, I think.
It will be imperative to optimize it.

I still have the idea of a timeline in my head.
Where the timeline would essentially be a list of shifts.
And the sequences would essentially just be a tuple of (indexFrom, indexTo), pointing to that list.
To retrieve the shifts, you then just do subList(indexFrom, indexTo), which is a no-op.

But I have no proof that it would actually be faster. (It could be *simpler*, though.)
It'd have to be tried.

Reply:
Intervals are currently not supported (nor do I think can be supported).
Consider the following example:

[1,2], [2,3], [3,4],[1,3],[3,5]

The following are all consecutive:
[1,2],[2,3],[3,4]
[1,3],[3,4]
[1,3],[3,5]
[2,3],[3,5]

In which case, what should the list of consecutive sequences be? There are duplicate nodes. Currently, what is supported is consecutive points, NOT consecutive intervals.

Reply:
I see your point, and that is why I estimated this to take longer than 1 sprint. :-) I believe we need to be able to answer the question, because this use case *will* show up.

In my opinion, overlapping intervals belong in the sequence.
It is the job of another constraint to make sure overlapping shifts do not happen.

Reply:
Perhaps `Sequence` is the wrong name here. `Cluster` sounds closer to what I think it should be.
It is all the shifts that happen between two breaks. If some of them overlap, it is no longer a sequence, it is a cluster.

Reply:
Added an interval based implementation for clusters in `IntervalTree`